### PR TITLE
Onedrive Business account permissions fix #633

### LINKED ISSUE #633: OneDrive: 'account loading...'
**Describe the bug:** just after rolling out this extension for more users, it seems there's an issue between Authenticator and OneDrive. The backup page is stuck on 'account - loading...'

Perhaps something changed on OneDrive, or perhaps it's a bug; how do we debug this?

**Platform:** Windows 10
 - Browser: Chrome, FireFox
 - Browser Version: latest versions

In Azure's audit logs, we do see a lot of "app role assignment grant to user" failures (Microsoft.Online.DirectoryServices.UniqueKeyPropertyException)

With my personal/private OneDrive account, it works as expected; but not within our corporate O365 environment.

---

After I explicitly sign out and then sign in again, this appears in the debugger:

Error 403 - XHR GET https://graph.microsoft.com/v1.0/me/drive/special/approot

Request

```
Accept-Encoding | gzip, deflate, br
-- | --
Accept-Language | en-US,en;q=0.5
Authorization | Bearer   <redacted>
Connection | keep-alive
Cookie | s_preRef=%5B%5B%27neowin.net%27%2C%271492531998951%27%5D%5D;  s_fid=1DB9DB9943B5A8EF-1329F236115FC0EE;  _ga=GA1.2.1340837329.1525429298; LPVID=BkZWMyMWJiOTZjMzdhZWE5;  mbox=PC#b4abd76be1114733b279653e440d54b7.37_0#1677239875\|session#e6381b68c4914f58b5d0e3e2fa11f6ce#1613996935;   MC1=GUID=e539776adb3b46ccb03fb00329f2a55d&HASH=e539&LV=202002&V=4&LU=1580719930235;  MUID=246626B83BF46FDE02132CD03FF46C26; MSCC=1614337854;  mslocale={'u':'nl-be'}
Host | graph.microsoft.com
```

Response
```
{
  "error": {
    "code": "accessDenied",
    "message": "Access denied",
    "innerError": {
      "date": "2021-03-24T10:08:39",
      "request-id": "b65d8e4a-064f-4465-a4ac-0d8ec86b1206",
      "client-request-id": "b65d8e4a-064f-4465-a4ac-0d8ec86b1206"
    }
  }
}
````

Potentially related:
* https://github.com/OneDrive/onedrive-api-docs/issues/682
* https://docs.microsoft.com/en-US/onedrive/developer/rest-api/concepts/direct-endpoint-differences?view=odsp-graph-online#permissions -> seems like there's a 2.0 version?
* https://docs.microsoft.com/en-us/graph/permissions-reference#remarks-5 -> see remark: The Files.ReadWrite.AppFolder delegated permission is only valid for personal accounts and is used for accessing the App Root special folder with the OneDrive Get special folder Microsoft Graph API.

### PR DESCRIPTION
**Pull request by mymindstorm**

### CODE REVIEW

**Comment by jbostoen** [comment]
Thank you very very much for integrating this!

Should there be a note about permissions to the organization's administrators? Or perhaps to the user telling them they might need to consult their admin?

**Comment by mymindstorm** [comment]
> Should there be a note about permissions to the organization's administrators?

Added

> Or perhaps to the user telling them they might need to consult their admin?

The consent screen will prompt users to request admin approval if they cannot authorize the app. At some point in the future, the consent screen will mark the integration as "untrustworthy" due us not being able to register it with a verified publisher (no legal entity to register as a MS Partner with).

image

File: src/background.ts [hunk 1]
```diff
@@ -331,7 +331,9 @@ function getBackupToken(service: string) {
       redirUrl = encodeURIComponent(chrome.identity.getRedirectURL());
       authUrl = `https://login.microsoftonline.com/common/oauth2/v2.0/authorize?client_id=${
         getCredentials().onedrive.client_id
-      }&response_type=code&redirect_uri=${redirUrl}&scope=https%3A%2F%2Fgraph.microsoft.com%2FFiles.ReadWrite.AppFolder%20https%3A%2F%2Fgraph.microsoft.com%2FUser.Read%20offline_access&response_mode=query&prompt=consent`;
+      }&response_type=code&redirect_uri=${redirUrl}&scope=https%3A%2F%2Fgraph.microsoft.com%2FFiles.ReadWrite${
+        localStorage.oneDriveBusiness === "false" ? ".AppFolder" : ""
```
Comment (Sneezry) [inline src/background.ts]: Shoud `localStorage.oneDriveBusiness !== "true"` make more sense here?

Comment (mymindstorm) [reply src/background.ts]: They both work, I just thought that this would be easier to read.

Comment (Sneezry) [reply src/background.ts]: My concern is if `localStorage.oneDriveBusiness` is `undefined`, this logic should be incorrect.

Comment (jbostoen) [reply src/background.ts]: I'm unfamiliar with extensions or the settings, but I was mostly wondering why there are quotes around the boolean?

Comment (Sneezry) [reply src/background.ts]: localStorage can only store string values

**Review by Sneezry — APPROVED** [review]

**Comment by jbostoen** [comment]
Any idea when this will result in a new release? 😁

**Comment by mymindstorm** [comment]
See #646

**Comment by Sneezry** [comment]
> Any idea when this will result in a new release? 😁

For Chrome, because we have more than 1.5M users, we need incremental rollouts. It could be as long as one month to deliver this version to all existing users. New-install users are able to install this version once it's available in CWS. This version should be available in CWS within one week.

For Firefox, @mymindstorm correct me if I have any mistake, it should be already available to all users, including existing users and new-install users.

For Edge, we may need some time before it is available to public, but it could not be too long.

**Comment by jbostoen** [comment]
Thanks @Sneezry !

Question: is the 'manual sync' actually a one way backup rather than a sync?

I set up the Authenticator with a OneDrive Business on 2 different computers. 
They both already had different TOTPs; in the JSON file on OneDrive Business I only see a backup of one of those 2 lists at a time.

**Comment by mymindstorm** [comment]
> For Firefox, @mymindstorm correct me if I have any mistake, it should be already available to all users, including existing users and new-install users.

It is usually immediately available. They _used_ to have incremental rollouts but removed it for some reason.

> For Edge, we may need some time before it is available to public, but it could not be too long.


Edge usually takes a week or two for certification and is available immediately. 


> Question: is the 'manual sync' actually a one way backup rather than a sync

We will have to correct the string. It is a one way backup and is referenced that way in all other parts of the ui. The only "sync" is the browsers built in extension data sync. I would be wary of using it without backups because it tends to be finicky.

**Comment by jbostoen** [comment]
Hm, thanks for the answer @mymindstorm . I'll have to discuss it with colleagues, but it it would be more clear to remove the word "sync" as it implies two-way sync

**Pull request merged**